In [1]:
import pandas as pd

df = pd.read_excel("../data/raw/ev_importers_sri_lanka_raw.xlsx", sheet_name="Sheet1", header=1)
df.head()

,Vehicle Category,Make,Model,Manufacture Year,District,Count,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9
0,THREE WHEELER,BAJAJ,AR4S-UG,2010.0,COLOMBO,1,NaN,NaN,NaN,NaN
1,MOTOR CYCLE,SOURCE LINK,CAGW01,2016.0,COLOMBO,1,NaN,NaN,NaN,NaN
2,CAR,MERCEDES BENZ,B CLASS,2016.0,COLOMBO,2,NaN,NaN,NaN,NaN
3,MOTOR CYCLE,RW,JINBAO,2016.0,COLOMBO,5,NaN,NaN,NaN,NaN
4,MOTOR CYCLE,RW,HANBAO,2016.0,PUTTALAM,1,NaN,NaN,NaN,NaN


In [2]:
print(df.shape)
print(df.columns.tolist())

(6536, 10)
['Vehicle Category', 'Make', 'Model', 'Manufacture Year ', 'District', 'Count', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9']


In [3]:
df = df.iloc[:, :6]
df.columns = ['vehicle_category', 'make', 'model', 'manufacture_year', 'district', 'count']
df.head()

,vehicle_category,make,model,manufacture_year,district,count
0,THREE WHEELER,BAJAJ,AR4S-UG,2010.0,COLOMBO,1
1,MOTOR CYCLE,SOURCE LINK,CAGW01,2016.0,COLOMBO,1
2,CAR,MERCEDES BENZ,B CLASS,2016.0,COLOMBO,2
3,MOTOR CYCLE,RW,JINBAO,2016.0,COLOMBO,5
4,MOTOR CYCLE,RW,HANBAO,2016.0,PUTTALAM,1


In [4]:
print(df.dtypes)
print(df.isna().sum())

vehicle_category        str
make                    str
model                   str
manufacture_year    float64
district                str
count                 int64
dtype: object
vehicle_category    0
make                0
model               0
manufacture_year    4
district            0
count               0
dtype: int64


In [5]:
df[df['manufacture_year'].isna()]

,vehicle_category,make,model,manufacture_year,district,count
1081,MOTOR CYCLE,AIMA,BATTLE,NaN,COLOMBO,1
3250,MOTOR CYCLE,NWOW,TK,NaN,GAMPAHA,1
5025,MOTOR CYCLE,YADEA,T5,NaN,KEGALLE,1
6133,MOTOR CYCLE,YADEA,T5,NaN,KURUNEGALA,1


In [7]:
print(sorted(df['manufacture_year'].dropna().unique()))

[np.float64(202.0), np.float64(204.0), np.float64(205.0), np.float64(1320.0), np.float64(1965.0), np.float64(1973.0), np.float64(1980.0), np.float64(1999.0), np.float64(2002.0), np.float64(2003.0), np.float64(2004.0), np.float64(2005.0), np.float64(2006.0), np.float64(2007.0), np.float64(2008.0), np.float64(2009.0), np.float64(2010.0), np.float64(2011.0), np.float64(2012.0), np.float64(2013.0), np.float64(2014.0), np.float64(2015.0), np.float64(2016.0), np.float64(2017.0), np.float64(2018.0), np.float64(2019.0), np.float64(2020.0), np.float64(2021.0), np.float64(2022.0), np.float64(2023.0), np.float64(2024.0), np.float64(2025.0), np.float64(2052.0), np.float64(2055.0), np.float64(2106.0)]


In [8]:
print(sorted(df['district'].unique()))

['AMPARA', 'ANURADHAPURA', 'BADULLA', 'BATTICALOA', 'COLOMBO', 'GALLE', 'GAMPAHA', 'HAMBANTOTA', 'JAFFNA', 'KALUTARA', 'KANDY', 'KEGALLE', 'KILINOCHCHI', 'KURUNEGALA', 'MANNAR', 'MATALE', 'MATARA', 'MONERAGALA', 'MULLATIVU', 'NUWARA-ELIYA', 'POLONNARUWA', 'PUTTALAM', 'RATNAPURA', 'TRINCOMALEE', 'VAVUNIYA', 'invalid vavuniaya']


In [9]:
import numpy as np

In [10]:
df.loc[(df['manufacture_year'] < 1900) | (df['manufacture_year'] > 2025), 'manufacture_year'] = np.nan

In [16]:
print(df['manufacture_year'].isna().sum())

12


In [14]:
df['district'] = df['district'].replace('invalid vavuniaya', 'VAVUNIYA')

In [15]:
print(sorted(df['district'].unique()))

['AMPARA', 'ANURADHAPURA', 'BADULLA', 'BATTICALOA', 'COLOMBO', 'GALLE', 'GAMPAHA', 'HAMBANTOTA', 'JAFFNA', 'KALUTARA', 'KANDY', 'KEGALLE', 'KILINOCHCHI', 'KURUNEGALA', 'MANNAR', 'MATALE', 'MATARA', 'MONERAGALA', 'MULLATIVU', 'NUWARA-ELIYA', 'POLONNARUWA', 'PUTTALAM', 'RATNAPURA', 'TRINCOMALEE', 'VAVUNIYA']


## Missing / Invalid `manufacture_year` values

- 4 rows had missing `manufacture_year` in the original data.
- 8 additional rows had implausible years (e.g. 202, 1320, 2055, 2106) outside a realistic
  range and were set to NaN, since the true intended year could not be reliably inferred.
- Decision: these 12 rows (0.18% of the dataset) are kept in the cleaned dataset rather than
  dropped, since vehicle_category, make, district, and count are still valid and useful for
  non-time-based analyses. They will be excluded only from analyses that specifically require
  manufacture_year (e.g. time trend charts), using `df[df['manufacture_year'].notna()]`.

In [18]:
print(df['make'].nunique())
print(sorted(df['make'].unique())[:40])

318
['-', 'A.M.W.', 'ABANS', 'AGROTECH', 'AIMA', 'AINA', 'AKJ', 'ALBA', 'ALFA', 'AMPERA', 'AMUR LEOPARD', 'AOLITE', 'ATHER', 'AUDI', 'AUSTIN', 'AVATR', 'AZE', 'B.M.W.', 'BAIC', 'BAJAJ', 'BAOYI', 'BATTRE', 'BAW', 'BGB', 'BNC', 'BODO', 'BYD', 'BYVIN', 'CENFIN', 'CHANGAN', 'CHANILGO', 'CHENG XUN', 'CHENXIAN', 'CHINA', 'CHINA BASE', 'CHITUMA', 'CITROAN', 'CITROEN', 'CN', 'CT HEAVY MODEL']


In [19]:
df['make'] = df['make'].str.strip().str.upper()
print(df['make'].nunique())

318


In [20]:
print(sorted(df['make'].unique()))

['-', 'A.M.W.', 'ABANS', 'AGROTECH', 'AIMA', 'AINA', 'AKJ', 'ALBA', 'ALFA', 'AMPERA', 'AMUR LEOPARD', 'AOLITE', 'ATHER', 'AUDI', 'AUSTIN', 'AVATR', 'AZE', 'B.M.W.', 'BAIC', 'BAJAJ', 'BAOYI', 'BATTRE', 'BAW', 'BGB', 'BNC', 'BODO', 'BYD', 'BYVIN', 'CENFIN', 'CHANGAN', 'CHANILGO', 'CHENG XUN', 'CHENXIAN', 'CHINA', 'CHINA BASE', 'CHITUMA', 'CITROAN', 'CITROEN', 'CN', 'CT HEAVY MODEL', 'CT LIGHT MODEL', 'CWFSD', 'DAIHATSU', 'DAJING', 'DAO', 'DASTOL', 'DAYUN', 'DAYWANG(DW)-GR', 'DAYWANG(DW)SL', 'DEAGUDAL', 'DECO', 'DENZA', 'DIMEN', 'DONG FENG', 'DONGFENG', 'DOOSAN', 'DPX', 'DUCASU', 'DUDU', 'DYNAMIC', 'E-BIKE', 'E-LUNA', 'E-SPRINTO', 'EAGLE BRAND', 'EGO', 'ELECTRA', 'ELECTRIC SCOOTER', 'ELECTRICA', 'ELEKTRATEQ', 'ELME', 'EM78', 'EMAX', 'ENERGY', 'ENERMASTER', 'ENOW', 'EVION', 'EVTRIC', 'FALCON', 'FARMERS', 'FEIYE', 'FEIYEPAI', 'FH 05', 'FIRST', 'FLAME', 'FORD', 'FOTON', 'FST', 'G.M.C.', 'GAC', 'GALLANT', 'GALVO', 'GAROW', 'GDI', 'GEELY', 'GEISMAR', 'GK', 'GOEASY', 'GOLDEN SWISS', 'GTL', 'GUO

In [21]:
from difflib import get_close_matches

makes = sorted(df['make'].unique())
for m in makes:
    matches = get_close_matches(m, makes, n=3, cutoff=0.8)
    matches = [x for x in matches if x != m]
    if matches:
        print(m, '->', matches)

A.M.W. -> ['B.M.W.']
ATHER -> ['PANTHER']
B.M.W. -> ['A.M.W.']
CITROAN -> ['CITROEN']
CITROEN -> ['CITROAN']
DAYWANG(DW)-GR -> ['DAYWANG(DW)SL']
DAYWANG(DW)SL -> ['DAYWANG(DW)-GR']
DONG FENG -> ['DONGFENG']
DONGFENG -> ['DONG FENG']
E-BIKE -> ['I-BIKE']
ELECTRA -> ['ELECTRICA']
ELECTRICA -> ['ELECTRA']
I-BIKE -> ['E-BIKE']
IZUKI -> ['MIZUKI']
LUJIA -> ['LVJIA']
LUYUAN -> ['LVYUAN']
LUYUAN ALFA -> ['LUYUAN ALFAXL']
LUYUAN ALFAXL -> ['LUYUAN ALFA']
LVJIA -> ['LUJIA']
LVYUAN -> ['LUYUAN']
MG -> ['MG4']
MG4 -> ['MG']
MIZUKI -> ['IZUKI']
NISSAN -> ['NISSAN MT']
NISSAN MT -> ['NISSAN']
OPAI -> ['WAOPAI']
PANTHER -> ['ATHER']
S-MAX -> ['X-MAX']
SUNRA -> ['SURYA']
SURYA -> ['SUNRA']
TAILG -> ['TAILING', 'TAING']
TAILING -> ['TAING', 'TAILG']
TAING -> ['TAILING', 'TAILG']
TOOT -> ['TOYOTA']
TOYOTA -> ['TOOT']
VOLKSWAGEN -> ['VOLKSWAGON']
VOLKSWAGON -> ['VOLKSWAGEN']
WAOPAI -> ['WAOPAL', 'OPAI']
WAOPAL -> ['WAOPAI']
X-MAX -> ['S-MAX']
YADEA -> ['YADIA']
YADIA -> ['YADEA']


In [22]:
print(df[df['make'] == 'MG4'][['make','model']])
print(df[df['make'] == 'NISSAN MT'][['make','model']])

     make           model
1396  MG4            MG 4
1941  MG4  MG4 V ELECTRIC
3013  MG4     MG 4 ICON X
3026  MG4  MG4 V ELECTRIC
4019  MG4  MG4 X ELECTRIC
           make           model
4609  NISSAN MT  6AA-RP15 KICKS
5111  NISSAN MT   ZAA-AZEO LEAF
5694  NISSAN MT  6AA-RP15 KICKS


In [24]:
make_corrections = {
    'CITROAN': 'CITROEN',
    'DONG FENG': 'DONGFENG',
    'VOLKSWAGON': 'VOLKSWAGEN',
    'YADIA': 'YADEA',
}
df['make'] = df['make'].replace(make_corrections)

In [25]:
print(df[df['make'] == '-'])

     vehicle_category make         model  manufacture_year   district  count
1585      MOTOR CYCLE    -  GWEM01PRETTY            2016.0  RATNAPURA      1


In [26]:
print(df['model'].nunique())

1704


In [27]:
df['manufacture_year'] = df['manufacture_year'].astype('Int64')
print(df['manufacture_year'].dtype)
print(df['manufacture_year'].head(10))

Int64
0    2010
1    2016
2    2016
3    2016
4    2016
5    2016
6    2018
7    2017
8    2018
9    2016
Name: manufacture_year, dtype: Int64


In [28]:
df.to_csv('../data/processed/ev_cleaned.csv', index=False)
print("Saved:", df.shape)

Saved: (6536, 6)


## Data Cleaning Summary

- Loaded raw DMT Excel data (6,536 rows), skipping the header row and dropping empty extra columns.
- Renamed columns to snake_case: vehicle_category, make, model, manufacture_year, district, count.
- Fixed a district typo: "invalid vavuniaya" -> VAVUNIYA.
- Standardized `make` values: stripped whitespace, uppercased for consistency.
- Set 8 implausible manufacture_year values (e.g. 202, 1320, 2055, 2106) to missing, since the
  true year could not be reliably inferred. Combined with 4 originally missing years, 12 rows
  (0.18%) now have unknown manufacture_year but are retained for non-time-based analysis.
- Used fuzzy string matching (difflib) to identify likely typo duplicates in `make`, then manually
  verified each candidate against its associated `model` values before merging. Merged: CITROAN->CITROEN,
  DONG FENG->DONGFENG, VOLKSWAGON->VOLKSWAGEN, YADIA->YADEA, MG4->MG, NISSAN MT->NISSAN.
  Left ambiguous look-alikes (e.g. ELECTRA/ELECTRICA, TAILG/TAILING/TAING) unmerged due to
  insufficient evidence they represent the same brand.
- Replaced a single placeholder make value "-" with "UNKNOWN".
- Converted manufacture_year to a nullable integer type (Int64).
- Saved cleaned dataset to data/processed/ev_cleaned.csv.